reinforcement_learning/bandits_recsys_movielens_testbed/src/train.py

```diff
@@ -0,0 +1,130 @@
+import json
+import logging
+import os
+
+from env import MovieLens100KEnv
+from io_utils import extract_model
+from vw_agent import VWAgent
+from vw_utils import MODEL_CHANNEL, MODEL_OUTPUT_DIR, DATA_OUTPUT_DIR
+
+logging.basicConfig(level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+
+def main():
+    """ Train a Vowpal Wabbit (VW) model through C++ process. """
+
+    channel_names = json.loads(os.environ['SM_CHANNELS'])
+    hyperparameters = json.loads(os.environ['SM_HPS'])
+
+    # Fetch algorithm hyperparameters
+    num_arms = int(hyperparameters.get("num_arms", 0))  # Used if arm features are not present
+    num_policies = int(hyperparameters.get("num_policies", 3))
+    exploration_policy = hyperparameters.get("exploration_policy", "egreedy").lower()
+    epsilon = float(hyperparameters.get("epsilon", 0))
+    mellowness = float(hyperparameters.get("mellowness", 0.01))
+    arm_features_present = bool(hyperparameters.get("arm_features", True))
+
+    # Fetch environment parameters
+    item_pool_size = int(hyperparameters.get("item_pool_size", 0))
+    top_k = int(hyperparameters.get("top_k", 5))
+    max_users = int(hyperparameters.get("max_users", 3))
+    total_interactions = int(hyperparameters.get("total_interactions", 1000))
+
+    if not arm_features_present and num_arms is 0:
+        raise ValueError("Customer Error: Please provide a non-zero value for 'num_arms'")
+    logging.info("channels %s" % channel_names)
+    logging.info("hps: %s" % hyperparameters)
+
+    # Different exploration policies in VW
+    # https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Contextual-Bandit-algorithms
+    valid_policies = ["egreedy", "bag", "regcbopt", "regcb"]
+    if exploration_policy not in valid_policies:
+        raise ValueError(f"Customer Error: exploration_policy must be one of {valid_policies}.")
+
+    if exploration_policy == "egreedy":
+        vw_args_base = f"--cb_explore_adf --cb_type mtr --epsilon {epsilon}"
+    elif exploration_policy in ["regcbopt", "regcb"]:
+        vw_args_base = f"--cb_explore_adf --cb_type mtr --{exploration_policy} --mellowness {mellowness}"
```

Comment:
does this mean the data provided can have labels?

Reply:
this mellowness is the c_0 parameter of the regcb algorithm